File: java/smallestSubtreeWithAllTheDeepestNodes.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    //hashmap maps values of nodes onto the max depth of its descendants
    private Map<Integer, Integer> valuesToMaxDepths = new HashMap();
    
    public TreeNode subtreeWithAllDeepest(TreeNode root) {
        //recursive visit function maps all node values onto max depth of their descendants 
        visit(0, root);
        //maxDepth is the max depth of the root and thus max depth of the tree
        int maxDepth = valuesToMaxDepths.get(root.val);
        TreeNode pointer = root;
        //start with the root; if a node has exactly one descendant or one descandant has a 
        //max depth lower than the tree's max depth, it's not the lowest common ancestor
        while (pointer.left == null || pointer.right == null || 
              valuesToMaxDepths.get(pointer.left.val) != maxDepth ||
              valuesToMaxDepths.get(pointer.right.val) != maxDepth) {
            //if a node is a leaf, it's the lowest common ancestor
            if (pointer.left == null && pointer.right == null) {
                return pointer;
            }
            //otherwise continue down the path
            if (pointer.left == null || valuesToMaxDepths.get(pointer.left.val) != maxDepth) {
                pointer = pointer.right;
            } else {
                pointer = pointer.left;
            }
        }
        return pointer;
    }
    
    private int visit(int level, TreeNode node) {
        //recursive function uses post order traversal to map node values onto max depths of descandant nodes
        int rightMaxDepth = node.right == null ? level : visit(level + 1, node.right);
        int leftMaxDepth = node.left == null ? level : visit(level + 1, node.left);
        int maxDepth = Math.max(rightMaxDepth, leftMaxDepth);
        valuesToMaxDepths.put(node.val, maxDepth);
        return maxDepth;
    }
}

File: python/smallest_subtree_with_all_the_deepest_nodes.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def subtreeWithAllDeepest(self, root: TreeNode) -> TreeNode:
        #the max_depths dictionary will map the value of a node to the greatest
        #depth of its descendants
        max_depths = dict()
        def visit(node: TreeNode, level: int) -> int:
            #determine if the node is a leaf:
            if node.left is None and node.right is None:
                max_depths[node.val] = level
                return level
            depths = []
            if node.left is not None:
                depths.append(visit(node.left, level + 1))
            if node.right is not None:
                depths.append(visit(node.right, level + 1))
            max_depth = max(depths)
            max_depths[node.val] = max_depth
            return max_depth
        visit(root, 0)
        #now traverse the tree until we find a node in which two children have the deepest max depth
        deepest = max_depths[root.val]
        node_pointer = root
        #if a node has exactly one descendant or one of its descendants doesn't contain the deepest level, it's not the lowest
        #common ancestor
        while node_pointer.left is None or node_pointer.right is None or max_depths[node_pointer.left.val] != deepest or max_depths[node_pointer.right.val] != deepest:
            #if the node is a leaf, then this is the lowest common ancestor
            if node_pointer.left is None and node_pointer.right is None:
                return node_pointer
            #otherwise, go down the tree
            elif node_pointer.left is None or (node_pointer.left is not None and max_depths[node_pointer.left.val]) != deepest:
                node_pointer = node_pointer.right
            else:
                node_pointer = node_pointer.left
        return node_pointer